Add Swarm management CLI commands

dnephin opened the pull request:
As described in our [ROADMAP.md](https://github.com/docker/docker/blob/master/ROADMAP.md#14-cluster-capable-engine), introduce new Swarm management commands to call to the corresponding API endpoints.

This PR is fully backward compatible (joining a Swarm is an optional feature of the Engine, and existing commands are not impacted)

dnephin committed:
Add Swarm management CLI commands

As described in our ROADMAP.md, introduce new Swarm management commands
to call to the corresponding API endpoints.

This PR is fully backward compatible (joining a Swarm is an optional
feature of the Engine, and existing commands are not impacted).

Signed-off-by: Daniel Nephin <dnephin@docker.com>
Signed-off-by: Victor Vieux <vieux@docker.com>
Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

amitshukla committed:
CLI docs

Signed-off-by: Amit Shukla <amit.shukla@docker.com>

aluzzardi commented on api/client/service/opts.go:
```diff
@@ -0,0 +1,372 @@
+	if !ok {
+		return fmt.Errorf("Failed to parse %v as a rational number", value)
+	}
+	*c = nanoCPUs(cpu.Mul(cpu, big.NewRat(1e9, 1)).Num().Int64())
```
Before calling `Int64()`, you have to check `IsInt()` or return an error. This is to make sure there's no loss (user input is not below nanos).

aluzzardi commented on api/client/service/opts.go:
```diff
@@ -0,0 +1,372 @@
+}
+
+func (m *memBytes) Set(value string) error {
+	val, err := units.RAMInBytes(value)
```
Check `err`

dnephin replied:
In what case would I need to check `err` here? We always return the err, so if there is an error it will be reported. 

This is how all the types in pflag are implemented, ex: https://github.com/spf13/pflag/blob/master/int64.go#L16-L19

aluzzardi replied:
Right - my bad!

dnephin committed:
New format for setting mounts on a service.

Signed-off-by: Daniel Nephin <dnephin@docker.com>

Commit:
Shorten `docker service` restart policy flags

Rename `--restart-policy-*` flags to `--restart-*` (e.g.,
`--restart-policy-delay` becomes `--restart-delay`).

Signed-off-by: Arnaud Porterie (icecrime) <arnaud.porterie@docker.com>

Commit:
Clarify services update policies

Rename `--updateconfig-*` to `--update-*` and improve help messages.

Signed-off-by: Arnaud Porterie (icecrime) <arnaud.porterie@docker.com>

vieux committed:
use ID for '*'

Signed-off-by: Victor Vieux <vieux@docker.com>

vieux committed:
add CACertHash: cli

Signed-off-by: Victor Vieux <vieux@docker.com>

vdemeester commented on api/client/node/inspect.go:
```diff
@@ -0,0 +1,141 @@
+func runInspect(dockerCli *client.DockerCli, opts inspectOptions) error {
+	client := dockerCli.Client()
+	getRef := func(ref string) (interface{}, []byte, error) {
+		ctx := context.Background()
```
hum, should `ctx` declared in the *upper* block (alongside `client`) ?

dnephin replied:
yup, fixed

vdemeester commented on api/client/node/list.go:
```diff
@@ -0,0 +1,123 @@
+	client := dockerCli.Client()
+
+	nodes, err := client.NodeList(
+		context.Background(),
```
The context should be shared here :

```go
ctx := context.Background()

nodes, err := client.NodeList(ctx, …)
// […]
info, err := client.Info(ctx)
// […]
```

vdemeester commented on api/client/node/remove.go:
```diff
@@ -0,0 +1,35 @@
+func runRemove(dockerCli *client.DockerCli, args []string) error {
+	client := dockerCli.Client()
+	for _, nodeID := range args {
+		err := client.NodeRemove(context.Background(), nodeID)
```
Here too, context should be shared.

```go
ctx := context.Background()
for _, nodeID := range args {
    err := client.NodeRemove(ctx, …)
    // […]
}
```

dnephin commented:
Ok, those issues should be fixed.

tonistiigi committed:
Fix accept policy initialization

- Set secret when using default policy
- Always initialize policy for all roles

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

tonistiigi committed:
Update join command message

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

dnephin committed:
Fix CLI side of volume problem.

Signed-off-by: Daniel Nephin <dnephin@docker.com>

vieux committed:
update to newest engine-api (cli)

Signed-off-by: Victor Vieux <vieux@docker.com>

Commit:
Rename `--port` to `--publish` in `docker service`

Rename `--port` to `--publish` for consistency between `docker service`
and `docker run`.

Signed-off-by: Arnaud Porterie (icecrime) <arnaud.porterie@docker.com>

tonistiigi committed:
Fix RenameCheckNames test

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

chanwit commented on api/client/node/promote.go:
```diff
@@ -0,0 +1,40 @@
+
+	cmd := &cobra.Command{
+		Use:   "promote NODE [NODE...]",
+		Short: "Promote a node as manager in the swarm",
```
Probably, "Promote a node to manager in the swarm"?

chanwit commented on api/client/node/demote.go:
```diff
@@ -0,0 +1,40 @@
+
+	cmd := &cobra.Command{
+		Use:   "demote NODE [NODE...]",
+		Short: "Demote a node as manager in the swarm",
```
Shoud it be "Demote a node from manager in the swarm" ?

vdemeester commented:
win2lin error is weird :
```
04:14:18 # github.com/docker/swarmkit/identity
04:14:18 vendor\src\github.com\docker\swarmkit\identity\randomid.go:54: nn.Text undefined (type big.Int has no field or method Text)
```

thaJeztah commented:
@vdemeester it's because that CI is still on Go 1.5.4

dnephin committed:
Fixes #273 - set command line args in args not command

Signed-off-by: Daniel Nephin <dnephin@docker.com>

dnephin committed:
backport fixes from docker/docker PR comments.

Signed-off-by: Daniel Nephin <dnephin@docker.com>

vieux committed:
upadate engine-api: cli

Signed-off-by: Victor Vieux <vieux@docker.com>

Commit:
Pretty print strings on human-readable outputs

Capitalize first letter in output meant for human consumptions (e.g.,
`docker service ls`, or `docker node inspect --pretty`).

Signed-off-by: Arnaud Porterie (icecrime) <arnaud.porterie@docker.com>

Commit:
Fix incorrect `docker service inspect` help message

Signed-off-by: Arnaud Porterie (icecrime) <arnaud.porterie@docker.com>

chanwit commented:
Just let you know, I did a rebase on master with a small conflict at:
` cli/usage.go`.

Other than that it's just fine.

vieux committed:
fix docker swarm inspect output on error

Signed-off-by: Victor Vieux <vieux@docker.com>

vieux committed:
git rm reject.go

Signed-off-by: Victor Vieux <vieux@docker.com>

vieux committed:
add -h back to inspect

Signed-off-by: Victor Vieux <vieux@docker.com>

vieux committed:
cli

Signed-off-by: Victor Vieux <vieux@docker.com>

vieux committed:
api

Signed-off-by: Victor Vieux <vieux@docker.com>

dnephin committed:
Upate from PR review.

Signed-off-by: Daniel Nephin <dnephin@docker.com>

mavenugo committed:
swarm managed networks should me marked as "swarm" scoped (cli)

Signed-off-by: Madhu Venugopal <madhu@docker.com>

Commit:
Add `docker service scale` command

Signed-off-by: Arnaud Porterie (icecrime) <arnaud.porterie@docker.com>

mrjana committed:
Vendoring libnetwork @892e1b9864d1f (cli)

Adds overlay datapath security enhancement
Adds netlink performance enhancements
Bug fixes

Signed-off-by: Jana Radhakrishnan <mrjana@docker.com>

dongluochen committed:
Upate auto-accept help text.

Signed-off-by: Dong Chen <dongluo.chen@docker.com>

dnephin committed:
Small UX improvements to 'docker service scale'

Signed-off-by: Daniel Nephin <dnephin@docker.com>

cpuguy83 committed:
Fix scale pointer value in service inspect -p

Before this change, pretty output shows `859533555312` instead of `1`.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

Commit:
Add missing newline

Signed-off-by: Arnaud Porterie (icecrime) <arnaud.porterie@docker.com>

dperny committed:
Fix command not displaying in service ls.

Closes #396.

Signed-off-by: Drew Erny <drew.erny@docker.com>

vieux committed:
fix rebase

Signed-off-by: Victor Vieux <vieux@docker.com>

icecrime commented:
LGTM :10

vieux commented:
LGTM 💯

tiborvass commented:
LGTM

icecrime commented:
CLI reference and docs will be completed in a followup :+1:

chanwit commented:
:tada :tada

I love the new `scale` command.
Thank you guys. This is super awesome!!

thaJeztah commented:
:bee: :bee: :tada:

albers commented:
@thaJeztah This PR added reference pages for the `swarm` and `node` commands. But there are none for `service` yet. Is there some WIP?

thaJeztah commented:
@albers correct, there's an issue for tracking this here; https://github.com/docker/docker/issues/23598

albers commented:
@thaJeztah cool, thanks. I already dropped a note there.
